odxtools/diaglayers/ecuvariant.py

```diff
@@ -0,0 +1,122 @@
+# SPDX-License-Identifier: MIT
+from copy import deepcopy
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Optional, Union, cast
+from xml.etree import ElementTree
+
+from typing_extensions import override
+
+from ..diagvariable import DiagVariable
+from ..dyndefinedspec import DynDefinedSpec
+from ..ecuvariantpattern import EcuVariantPattern
+from ..exceptions import odxassert
+from ..nameditemlist import NamedItemList
+from ..odxlink import OdxDocFragment, OdxLinkDatabase, OdxLinkRef
+from ..parentref import ParentRef
+from ..variablegroup import VariableGroup
+from .diaglayer import DiagLayer
+from .ecuvariantraw import EcuVariantRaw
+from .hierarchyelement import HierarchyElement
+
+
+@dataclass
+class EcuVariant(HierarchyElement):
+
+    @property
+    def ecu_variant_raw(self) -> EcuVariantRaw:
+        return cast(EcuVariantRaw, self.diag_layer_raw)
+
+    @property
+    def diag_variables_raw(self) -> List[Union[DiagVariable, OdxLinkRef]]:
+        return self.ecu_variant_raw.diag_variables_raw
+
+    @property
+    def dyn_defined_spec(self) -> Optional[DynDefinedSpec]:
+        return self.ecu_variant_raw.dyn_defined_spec
+
+    @property
+    def parent_refs(self) -> List[ParentRef]:
+        return self.ecu_variant_raw.parent_refs
+
+    @property
+    def ecu_variant_patterns(self) -> List[EcuVariantPattern]:
+        return self.ecu_variant_raw.ecu_variant_patterns
+
+    #######
+    # <properties subject to value inheritance>
+    #######
+    @property
+    def diag_variables(self) -> NamedItemList[DiagVariable]:
+        return self._diag_variables
+
+    @property
+    def variable_groups(self) -> NamedItemList[VariableGroup]:
```

Comment:
those are exclusive to ecu variants?

Reply:
nope, but the XSD defines them separately for ECU-SHARED-DATA, FUNCTIONAL-GROUP, BASE-VARIANT and ECU-VARIANT layers (note: not for PROTOCOL). To avoid confusion, I decided to simply match the spec even if I think that their approach is rather inelegant...

Reply:
python allows multi inheritance, so move them to shared class

Reply:
I think that this would be quite confusing for people who try to match the XSD with the python code. (if the specification did it this way, I would be all for it...)